resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -301,21 +317,53 @@ void add(DefaultDnsCacheEntry e) {
                                 assert replacedEntry == null;
                                 replacedEntry = entry;
                             }
-                        }
+                            i++;
+                        } while (i < entries.size());
                         newEntries.add(e);
-                        if (compareAndSet(entries, newEntries)) {
+                        if (compareAndSet(entries, Collections.unmodifiableList(newEntries))) {
                             if (replacedEntry != null) {
-                                replacedEntry.cancelExpiration();
+                                // We don't care if the replaced entry may have had a smaller TTL. At worse we will
+                                // clear the cache a bit to early which is fine as the TTL is just the maximum time
+                                // we are allowed to cache the entry.
                             }
+
+                            scheduleCacheExpirationIfNeeded(ttl, loop);
                             return;
                         }
                     } else if (compareAndSet(entries, Collections.singletonList(e))) {
+                        scheduleCacheExpirationIfNeeded(ttl, loop);
                         return;
                     }
                 }
             } else {
-                List<DefaultDnsCacheEntry> entries = getAndSet(Collections.singletonList(e));
-                cancelExpiration(entries);
+                set(Collections.singletonList(e));
+                scheduleCacheExpirationIfNeeded(ttl, loop);
+            }
+        }
+
+        private void scheduleCacheExpirationIfNeeded(int ttl, EventLoop loop) {
+            for (;;) {
+                ScheduledFuture<?> oldFuture = FUTURE_UPDATER.get(this);
+                if (oldFuture == null || oldFuture.getDelay(TimeUnit.SECONDS) > ttl) {
+                    ScheduledFuture<?> newFuture = loop.schedule(this, ttl, TimeUnit.SECONDS);
```

Comment:
dont want to be too nitpicky, but this isn't quite correct.  I think you need to calculate System.nanoTime() as the start point outside of the loop, and deduct when scheduling:

```java
long start = System.nanoTime();
long ttlNanos = TimeUnit.SECONDS.toNanos(ttl);
// ...
for (;;) {
  loop.schedule(this, ttlNanos - (System.nanoTime() - start), TimeUnit.NANOSECONDS);
}

```

Reply:
I think you are right in general I am just not sure we need to be so picky here... that said I guess it can’t really harm